ring_benchmark/ring.erl

```diff
@@ -0,0 +1,56 @@
+-module(ring).
+-export([main/1]).
+
+node_setup() -> node_setup(#{}).
+node_setup(State) ->
+    receive
+        {notify_time, Pid, MessageCount} ->
+            start_root_node(State#{pid => Pid, m => MessageCount});
+        {next_node, NextNode} ->
+            node_setup(State#{next_node => NextNode});
+        {send_message} ->
+            loop_normal_node(State)
+    end.
+
+start_root_node(State = #{next_node := NextNode}) ->
+    Now = erlang:timestamp(),
+    NextNode ! {send_message},
+    loop_root_node(State#{starting_time => Now}).
+
+loop_root_node(#{starting_time := StartingTime,
+                 pid := NotificationPid, m := 0}) ->
+    Now = erlang:timestamp(),
+    TimeDiffMs = timer:now_diff(Now, StartingTime),
+    NotificationPid ! {time, TimeDiffMs};
+
+loop_root_node(State = #{next_node := NextNode, m := MessageCount}) ->
+    NextNode ! {send_message},
+    receive
+        {send_message} -> loop_root_node(State#{m => MessageCount - 1})
+    end.
+
+loop_normal_node(State = #{next_node := NextNode}) ->
+    NextNode ! {send_message},
+    receive
+        {send_message} ->            
+            loop_normal_node(State)
+    end.
+
+% N processes, M messages
+main([NArg,MArg]) ->
+    N = arg_to_number(NArg),
+    M = arg_to_number(MArg),
+    [Root | Rest] = [spawn_link(fun () -> node_setup() end)
+                     || _Time <- lists:seq(1, N)],
+    [Pid1 ! {next_node, Pid2}
```

Comment:
You could simplify this exercise a lot by building a single list of Pids at the beginning to account for all the runs through the ring. For example for 3x2:

[Pid1, Pid2, Pid3, Pid1, Pid2, Pid3]

The main function would just pass that initial list and each node would just remove itself from the list and call the next process passing the new list. This all the processes run the same code, you don't need internal state and you can do all the time operations in main.

Reply:
I don't know if that's a valid solution for the original problem, but it sounds better. 

@igaray @unbalancedparentheses opinions?